src/app/ReadClient.h

```diff
@@ -315,6 +332,32 @@ class ReadClient : public Messaging::ExchangeDelegate
     // that's the only case when the consumer moved a ReadParams into the client.
     CHIP_ERROR SendAutoResubscribeRequest(ReadPrepareParams && aReadPrepareParams);
 
+    //
+    // This provides a standard re-subscription policy implementation that given a termination cause, does the following:
+    //      - Calculates the time till next subscription with fibonacci back-off (implemented by ComputeTimeTillNextSubscription()).
+    //      - Schedules the next subscription attempt at the computed interval from the previous step. Operational discovery and
+    //        CASE establishment will be attempted if aTerminationCause was CHIP_ERROR_TIMEOUT. In all other cases, it will attempt
+    //        to re-use a previously established session.
+    //
+    CHIP_ERROR DefaultResubscribePolicy(CHIP_ERROR aTerminationCause);
+
+    //
+    // Computes the time till the next re-subscription with millisecond resolution over
+    // an even increasing window following a fibonacci sequence with the current retry count
+    // used as input to the fibonacci algorithm.
+    //
+    // CHIP_RESUBSCRIBE_MAX_FIBONACCI_STEP_INDEX is used as the maximum ceiling for that input.
+    //
+    uint32_t ComputeTimeTillNextSubscription();
+
+    //
+    // Schedules a re-subscription aTimeTillNextResubscriptionMs into the future.
+    //
+    // If restablishCASE is true, operational discovery and CASE will be attempted at that time before
+    // the actual IM interaction is initiated.
+    //
+    CHIP_ERROR ScheduleResubscription(uint32_t aTimeTillNextResubscriptionMs, bool restablishCASE = true);
```

Comment:
I realize that this API doesn't actually accept a new SessionHandle as a possible argument in the case where the application undertook the responsibility of setting up the CASE session itself. Will likely amend the API to:

```
CHIP_ERROR ScheduleResubscription(uint32_t aTimeTillNextResubscriptionMs, Optional<SessionHandle> newSessionHandle, bool restablishCASE = true);
```

If `newSessionHandle` is present, that will be used and the value of `restablishCASE` will be ignored. If it isn't present, then restablishCASE will be assessed.
```

Reply:
See above about how the CASE re-establishment should just happen in the Callback, not here...

Reply:
The proposal to pass a `newSessionHandle` here triggered questions for me on the intended API flow for `ReadClient` related to peer binding.  Is this the following correct?

- `ReadClient` tracks subscriptions between a client and exactly one peer (implied by the fact that the class has a singular `mPeer` instance data.)
- The peer that a `ReadClient` object is bound to is set not in the constructor, but when `ReadClient::SendRequest` is called, where the peer info is passed via `ScheduleResubscription.mSessionHolder`.

Given this, shouldn't we strive to focus the external API more on the peer relationship, and setting the peer identity to track, rather than providing a particular active session (that could go away or be replaced) to that peer? 

Perhaps APIs like `SessionManager::FindSecureSessionForNode` could be leveraged to lookup the best known session for the bound peer (and trigger auto-creation if none exist) rather than have the app pass SessionHandles into more and more of the API surface?

Reply:
> Given this, shouldn't we strive to focus the external API more on the peer relationship, and setting the peer identity to track, rather than providing a particular active session (that could go away or be replaced) to that peer?

All the IM client-side objects only accept `SessionHandle`s today, so we have to keep with that for now for 1.0, since doing anything different is a much bigger change than we can swallow right now.